### PR TITLE
Rigol dg1062

### PR DESCRIPTION
**Pull request by sohailc**
A driver for the Rigol DG1062

### CODE REVIEW

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            val_mapping={1: 'POSITIVE', 0: 'NEGATIVE'},
+        )
+
+    def apply(self, **kwargs: Dict) ->Union[None, Dict]:
+        """
+        Apply a waveform on the channel
+
+        Example:
+        >>> gd = DG1062("gd", "TCPIP0::169.254.187.99::inst0::INSTR")
+        >>> gd.channels[0].apply(waveform="SIN", freq=1E3, ampl=1.0, offset=0, phase=0)
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/rigol/DG1062.py]: Is there a good way to know the valid arguments. Perhaps even just a ref to the relevant page in a manual

Comment (sohailc) [reply qcodes/instrument_drivers/rigol/DG1062.py]: The valid arguments depend on the waveform, which makes this a bit tricky. If you give wrong arguments I raise a ValueError and list the arguments expected. See the example notebook. 

Perhaps I can build an alternate interface which looks something like: 
```python
gd.channel[0].apply.sin(freq, ampl, offset, phase)
```
The advantage of that is that we can then have proper doc strings which the user can refer to.

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        """
+        Args:
+            parent: The instrument this channel belongs to
+            name (str)
+            channel (int)
+        """
+
+        super().__init__(parent, name)
+
+        self.parent = parent
```
[the remark this answers is not in the document]
Comment (sohailc) [reply qcodes/instrument_drivers/rigol/DG1062.py]: Of course, you can remove line 29.

[the remark this answers is not in the document]
Comment (sohailc) [reply qcodes/instrument_drivers/rigol/DG1062.py]: Right now it seems we do need both lines. Line 27 does not automatically add parent to self.

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 1 above)
Comment (sohailc) [reply qcodes/instrument_drivers/rigol/DG1062.py]: image

**Comment by sohailc** [comment]
@jenshnielsen can you have another look at this, please?

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 1 above)
Comment (astafan8) [reply qcodes/instrument_drivers/rigol/DG1062.py]: sweet!

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+from functools import partial
+from typing import Union, Dict
+
+from qcodes import VisaInstrument, validators as vals
+from qcodes import InstrumentChannel, ChannelList
+
+log = logging.getLogger(__name__)
+
+
+def sane_partial(func, docstring, **kwargs):
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: shall this reside in "helpers" somewhere?

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+            "on",
+            get_cmd=f":SOUR{channel}:BURS?",
+            set_cmd=f":SOUR{channel}:BURS {{}}",
+            vals=vals.Enum(0, 1, "ON", "OFF")
+        )
+
+        self.add_parameter(
+            "polarity",
+            get_cmd=f":SOUR{channel}:BURS:GATE:POL?",
+            set_cmd=f":SOUR{channel}:BURS:GATE:POL {{}}",
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: just for my understanding, why are double brackets needed here?

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+            set_cmd=f":SOUR{channel}:BURST:IDLE {{}}",
+            vals=vals.MultiType(
+                vals.Enum("FPT", "TOP", "BOTTOM", "CENTER"),
+                vals.Numbers()  # DIY
+            )
+        )
+
+    def trigger(self) ->None:
+        """
+        Send a soft trigger to the instrument. This only works if the trigger
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: "soft" means "from software"/"from the driver"?

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 3 above)
Comment (sohailc) [reply qcodes/instrument_drivers/rigol/DG1062.py]: Yeah that could be helpful. Where exactly would you propose to put this?

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 5 above)
Comment (sohailc) [reply qcodes/instrument_drivers/rigol/DG1062.py]: Yes. I will make this more clear

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 4 above)
Comment (astafan8) [reply qcodes/instrument_drivers/rigol/DG1062.py]: that i know, just haven't seen visa commands which contain brackets for the values ;)

File: qcodes/instrument_drivers/rigol/DG1062.py (hunk 3 above)
Comment (astafan8) [reply qcodes/instrument_drivers/rigol/DG1062.py]: `qcodes/utils/helpers.py` should be an OK place. Although that file is already huge...

File: qcodes/utils/helpers.py [hunk 6]
```diff
@@ -581,3 +582,32 @@ def attribute_set_to(object_: Any, attribute_name: str, new_value: Any):
     setattr(object_, attribute_name, new_value)
     yield
     setattr(object_, attribute_name, old_value)
+
+
+def sane_partial(func, docstring, **kwargs):
```
Comment (jenshnielsen) [inline qcodes/utils/helpers.py]: Could you give this a more descriptive name. Sane_partial tells very little about why this is more sane

**Comment by jenshnielsen** [comment]
@sohailc Looks good but I think we should find a better name for sane_partial

**Comment by WilliamHPNielsen** [comment]
`partial_with_docstring`? That's at least immediately understandable.

**Pull request merged**